### PR TITLE
Simplify scripts

### LINKED ISSUE #10160: Simplify yarn scripts for running e2e
The package.json scripts (`package.json` and `projects/storefrontapp-e2e-cypress/package.json`) contain plenty of different scripts for running e2e. They are almost duplicate with small differences, depending on many factors. It's hard to maintain. The scripts also have inconsistent naming (sometimes factors are separated with dashes, sometimes colons; the order of factors is inconsistent).

It would be good to have one (or at most few) simple basic commands for running e2e and everything else should be driven with dynamic arguments (i.e. preceded with `--`).

### List of identified factors:

| factor | possible values |
| - | - |
| backend version | <ul><li>`2005`</li><li>`1905` (should be dropped)</li> <li>`2011` (not added yet)</li> <li>`ccv2` (not sure if still needed)</li></ul> It's then mapped to specific OCC endpint prefix and in specific backend URL in a  relevant configuration file `cypress<2005/1905/ccv2>.json`  |
| scope | <ul><li>all (default)</li><li>`regression` only </li> <li>`smoke` only </li><li>`mobile` only </li><li>`acessibility` only</li><li>vendor only (covers CDS lib at the time of writing)</li></ul> |
| how to run | <ul><li> `run` (run in cli)</li><li> `open` (open in window)</li><li> `start-run` (start app and run e2e in cli)</li><li> `start-open` (start app and open in window)</li></ul>  Note: In CI now we use `start-run`. |
| storefront channel | <ul><li>b2c (default)</li><li>`b2b`</li></ul> | 
| where meant to run | <ul><li>locally</li><li>in CI -  needs `--record` arg and `--key` (secret password stored in CI) to push test results to Cypress Dashboard;</li></ul> |

Moreover, when runs are made on CI and results pushed to Cypress Dashboard, we'd like to add _cypress tags_ with used factors - so the test results in Cypress Dashboard can be eaisly categorized, identified and filtered.

### Other issues
Some rules for scope are overcomplicated now (i.e. include `regex` for running everything but not vendor or b2b: `--spec \"cypress/integration/!(vendor|b2b)/**/*.e2e-spec.ts\""`)

### Comments
When the factors need to be typed as arguments (`--`), the overall command will be longer and harder to remember. it would be good to provide at least some doc / comment with all possible arguments and values. Maybe we can encapsulate it in a separate sh script (with `--help` option).

### PR DESCRIPTION
**Pull request by hackergil**
Fixes #10160

- Spartacus variables now managed using `env-cmd`
- Env file has variables for different environments
- Removed a lot of targets that weren't being used
- Variables for bash (SPA_ENV, CYPRESS_CONFIG) set on CI (for devs, they'll need to be set locally)

### CODE REVIEW

File: extra-webpack.config.js [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
+  env[key] = value === 'true' ? true : value === 'false' ? false : value;
+});
 
 console.log('env=', env);
 
 module.exports = {
   plugins: [
     new webpack.DefinePlugin({
-      'build.process.env': env,
+      'buildProcess.env': JSON.stringify(env),
```
Comment (marlass) [inline extra-webpack.config.js:21]: What's the benefit of JSON.stringify here?

File: ci-scripts/e2e-cypress.sh [hunk 2]
```diff
@@ -18,6 +18,7 @@ do
     case "$1" in
         '--suite' | '-s' )
             SUITE=":$2"
+            ENV_FILE="-b2b"
```
Comment (marlass) [inline ci-scripts/e2e-cypress.sh]: Where this variable is used?

File: projects/storefrontapp/src/environments/environment.ts [hunk 3]
```diff
@@ -8,11 +8,10 @@ import { Environment } from './models/environment.model';
...
-    build.process.env.SPARTACUS_BASE_URL ??
-    'https://spartacus-devci767.eastus.cloudapp.azure.com:9002',
-  // 'https://spartacus-dev0.eastus.cloudapp.azure.com:9002',
-  occApiPrefix: build.process.env.SPARTACUS_API_PREFIX ?? '/occ/v2/',
-  cds: build.process.env.SPARTACUS_CDS ?? false,
-  b2b: build.process.env.SPARTACUS_B2B ?? false,
-  cdc: build.process.env.SPARTACUS_CDC ?? false,
+    buildProcess.env.CX_BASE_URL ??
+    'https://spartacus-dev0.eastus.cloudapp.azure.com:9002',
+  occApiPrefix: '/occ/v2/',
```
Comment (marlass) [inline projects/storefrontapp/src/environments/environment.ts:13]: If we no longer need to test 1905 why not to inline it in the app?

File: .env-cmdrc [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
+  "dev:b2c": {
```
Comment (marlass) [inline .env-cmdrc]: What's the benefit of grouping it like this vs having 4 environments: `dev, ci, b2b, b2c` and composing them?

File: package.json [hunk 5]
```diff
@@ -2,55 +2,48 @@
   "name": "storefrontapp",
   "version": "3.0.0",
   "license": "Apache-2.0",
+  "author": "SAP, Spartacus team",
   "engines": {
     "node": ">=10 <14"
   },
   "scripts": {
-    "build": "ng build storefrontapp --prod",
+    "build": "env-cmd -e ci:b2c,$SPA_ENV ng build storefrontapp",
```
Comment (marlass) [inline package.json]: Do we need non-prod build of the app?

File: package.json [hunk 6]
```diff
@@ -2,55 +2,48 @@
...
-    "start": "ng serve",
-    "start:ci:1905": "cross-env SPARTACUS_BASE_URL=https://spartacus-legacy.eastus.cloudapp.azure.com:9002 SPARTACUS_API_PREFIX=/rest/v2/ yarn start:prod",
-    "start:ci:2005": "cross-env SPARTACUS_BASE_URL=https://spartacus-devci767.eastus.cloudapp.azure.com:9002 SPARTACUS_API_PREFIX=/occ/v2/ yarn start:prod",
-    "build:ci:2005": "cross-env SPARTACUS_BASE_URL=https://spartacus-devci767.eastus.cloudapp.azure.com:9002 SPARTACUS_API_PREFIX=/occ/v2/ yarn build --prod",
-    "start:ci:ccv2": "cross-env SPARTACUS_BASE_URL=https://spartacus-demo.eastus.cloudapp.azure.com:8443 yarn start:prod",
-    "start:prod": "ng serve --prod",
-    "start:ssl": "ng serve --ssl",
+    "start": "env-cmd -e ci:b2c,$SPA_ENV ng serve",
+    "start:prod": "env-cmd -e ci:b2c,$SPA_ENV ng serve --prod",
+    "start:ssl": "env-cmd -e ci:b2c,$SPA_ENV ng serve --ssl",
```
Comment (marlass) [inline package.json]: Can we drop this one? I don't remember ever using it
```suggestion
```

File: package.json [hunk 7]
```diff
@@ -2,55 +2,48 @@
...
+    "e2e:run:ci": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci",
+    "e2e:run:ci:b2b": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:b2b",
+    "e2e:run:ci:flaky": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:flaky",
+    "e2e:run:ci:cds": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:cds",
+    "e2e:start:open": "start-server-and-test start http-get://localhost:4200 e2e:open",
+    "e2e:start:run:ci": "start-server-and-test start:prod http-get://localhost:4200 e2e:run:ci",
+    "e2e:start:run:ci:b2b": "start-server-and-test start:prod http-get://localhost:4200 e2e:run:ci:b2b",
+    "e2e:start:run:ci:cds": "start-server-and-test start:prod http-get://localhost:4200 e2e:run:ci:cds",
+    "e2e:start:run:ci:flaky": "start-server-and-test start:prod  http-get://localhost:4200 e2e:run:ci:flaky",
+    "e2e:start:open:cds": "start-server-and-test start:prod http-get://localhost:4200 e2e:open",
```
Comment (marlass) [inline package.json]: These cds scripts doesn't differ from non cds ones.

File: ci-scripts/e2e-cypress.sh [hunk 8]
```diff
@@ -73,6 +70,6 @@ fi
 echo '-----'
 echo "Running Cypress end to end tests"
 
-yarn e2e:cy"${INTEGRATION}":start-run-ci"${CI_ENV}${SUITE}"
+yarn e2e${INTEGRATION}:start:run:ci"${SUITE}"
```
Comment (marlass) [inline ci-scripts/e2e-cypress.sh]: Isn't suite and integration treated the same now? Only last part of the command?

**Review by marlass — CHANGES_REQUESTED** [review]

File: .env-cmdrc (hunk 4 above)
Comment (hackergil) [reply .env-cmdrc]: That we need combinations (different URLs for local and CI, set the B2B var to true). We don't have a `b2b` env alone.

File: ci-scripts/e2e-cypress.sh (hunk 2 above)
Comment (hackergil) [reply ci-scripts/e2e-cypress.sh]: It's not. Removed.

File: extra-webpack.config.js (hunk 1 above)
Comment (hackergil) [reply extra-webpack.config.js:21]: If it is not used, variable parsing is not done correctly for webpack

File: package.json (hunk 5 above)
Comment (hackergil) [reply package.json]: Nope. Thanks for catching this one

File: package.json (hunk 7 above)
Comment (hackergil) [reply package.json]: True. Removed.

File: package.json (hunk 6 above)
Comment (hackergil) [reply package.json]: Yes. Dropped.

File: projects/storefrontapp/src/environments/environment.ts (hunk 3 above)
Comment (hackergil) [reply projects/storefrontapp/src/environments/environment.ts:13]: We will need 2011 soon.

File: ci-scripts/e2e-cypress.sh (hunk 8 above)
Comment (hackergil) [reply ci-scripts/e2e-cypress.sh]: Yup. Integration removed.

File: .env-cmdrc (hunk 4 above)
Comment (marlass) [reply .env-cmdrc]: But this package allows to use different environments at the same time.

eg: `env-cmd ci,b2b <ommand>`

We have fallbacks in our environment.ts file and also composing will remove some duplication from this file. We would have single url for ci, etc.

File: projects/storefrontapp/src/environments/environment.ts (hunk 3 above)
Comment (marlass) [reply projects/storefrontapp/src/environments/environment.ts:13]: We will have different api prefix in 2011?

File: .github/workflows/flaky-e2e-tests.yml [hunk 9]
```diff
@@ -24,7 +24,7 @@ jobs:
         with:
           working-directory: projects/storefrontapp-e2e-cypress
       - name: Build Spartacus
-        run: yarn build:libs && yarn build:ci:2005
+        run: yarn build:libs && yarn build:prod
```
Comment (marlass) [inline .github/workflows/flaky-e2e-tests.yml]:
```suggestion
        run: yarn build:libs && yarn build
```

File: package.json [hunk 10]
```diff
@@ -2,55 +2,45 @@
...
-    "build": "ng build storefrontapp --prod",
+    "build": "env-cmd -e ci:b2c,$SPA_ENV ng build storefrontapp --prod",
     "build:assets": "yarn --cwd ./projects/assets build",
-    "build:libs": "ng build core --prod && ng build storefrontlib --prod && yarn build:schematics && yarn build:organization && yarn build:product && yarn build:storefinder && yarn build:qualtrics && yarn build:assets && yarn build:incubator && yarn build:cdc && yarn build:cds:lib && yarn build:setup",
+    "build:cdc": "ng build cdc --prod",
+    "build:cds": "ng build cds --prod",
     "build:incubator": "ng build incubator --prod",
+    "build:libs": "ng build core --prod && ng build storefrontlib --prod && yarn build:schematics && yarn build:organization && yarn build:product && yarn build:storefinder && yarn build:qualtrics && yarn build:assets && yarn build:incubator && yarn build:cdc && yarn build:cds && yarn build:setup",
+    "build:organization": "yarn --cwd feature-libs/organization run build:schematics && ng build organization --prod",
+    "build:prod": "env-cmd -e ci:b2c,$SPA_ENV ng build storefrontapp --prod",
```
Comment (marlass) [inline package.json]:
```suggestion
```

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/storefrontapp/src/environments/environment.ts (hunk 3 above)
Comment (hackergil) [reply projects/storefrontapp/src/environments/environment.ts:13]: ah, I see what you mean. Nope, we could inline in the app but I would keep it like this for this PR and just create a new ticket for that. What do you think?

File: .env-cmdrc (hunk 4 above)
Comment (hackergil) [reply .env-cmdrc]: the idea for the `dev` section is that devs can tweak it the way they want. For CI, I still think we need 2. Maybe we can discuss?

File: package.json (hunk 10 above)
Comment (hackergil) [reply package.json]: Incorporated

File: .github/workflows/flaky-e2e-tests.yml (hunk 9 above)
Comment (hackergil) [reply .github/workflows/flaky-e2e-tests.yml]: Done.

File: projects/storefrontapp/src/environments/environment.ts (hunk 3 above)
Comment (marlass) [reply projects/storefrontapp/src/environments/environment.ts:13]: If you prefer new ticket, sure. Let's do this separately.

File: .env-cmdrc (hunk 4 above)
Comment (marlass) [reply .env-cmdrc]: Yep. Let's discuss

**Review by marlass — DISMISSED** [review]

**Review by marlass — APPROVED** [review]

**Pull request merged**